[nativert] ensure that moveable outputs are set in other executionframe ctor

dolpm opened the pull request:
Summary:
so we use this constructor in HigherOrderKernel. problems arise in the loop condition, where it's possible for an output from the prev. iteration to be an input to the next. so the Output(N) of a kernel may be the Input(M) to a kernel in the next iteration. Thus, if the output value is reset (via. fastresizetozero) or overwritten by a prev. kernel before it is to be used, we have major major issues.

we need to enforce that outputs are moved, not copied, to ensure this doesn't happen.

Test Plan:
buck2 test //caffe2/test:test_export --local-only -- test_while_loop_tensor_constant_idx_cpp_runtime_nonstrict

Rollback Plan:

Differential Revision: D80565374

SherlockNoMad commented on torch/nativert/executor/ExecutionFrame.cpp:
```diff
@@ -169,6 +173,7 @@ std::vector<c10::IValue> ExecutionFrame::tryMoveUserOutputs() {
   for (const auto& [i, outputValue] : c10::enumerate(outputs)) {
     if (const Value* const* valuePtr = std::get_if<Value*>(&outputValue);
         valuePtr && *valuePtr) {
+      LOG(INFO) << "moveable: " << isOutputMovable(i);
```
clean up debug logging?

SherlockNoMad commented on torch/nativert/executor/ExecutionFrame.cpp:
```diff
@@ -125,6 +126,9 @@ void ExecutionFrame::updateMovableOutputs() {
       */
 
+      LOG(INFO) << "is moveable: " << persistent_[id] << " "
+                << (inputs.find(id) != inputs.end());
```
clean up?

If you really need it, make it VLOG

dolpm committed:
[nativert] ensure that moveable outputs are set in other executionframe ctor (#161005)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/161005

so we use this constructor in HigherOrderKernel. problems arise in the loop condition, where it's possible for an output from the prev. iteration to be an input to the next. so the Output(N) of a kernel may be the Input(M) to a kernel in the next iteration. Thus, if the output value is reset (via. fastresizetozero) or overwritten by a prev. kernel before it is to be used, we have major major issues.

we need to enforce that outputs are moved, not copied, to ensure this doesn't happen.

Test Plan:
buck2 test //caffe2/test:test_export --local-only -- test_while_loop_tensor_constant_idx_cpp_runtime_nonstrict

Rollback Plan:

Reviewed By: SherlockNoMad

Differential Revision: D80565374
